src/verifier/verifier.rs

```diff
@@ -162,19 +160,40 @@ impl WhammVisitor<Option<DataType>> for TypeChecker<'_> {
         // type check body
 
         self.table.enter_named_scope(&function.name.name);
-        self.visit_block(&function.body);
+        let check_ret_type = self.visit_block(&function.body);
         let _ = self.table.exit_scope();
 
-        // return type
-        todo!();
+        //figure out how to deal with void functions (return type is ())
+        if check_ret_type != Some(DataType::AssumeGood) || check_ret_type != function.return_ty {
+            self.err.type_check_error(
+                false,
+                format! {"Return type of function {} does not match the return type specified in the function signature", function.name.name},
+                &function.name.loc.clone().map(|l| l.line_col),
+            );
+        }
+        //return the type of the fn
+        function.return_ty.clone()
     }
 
     fn visit_block(&mut self, block: &Block) -> Option<DataType> {
-        // TODO: finish user def function type checking
+        let mut ret_type = None;
         for stmt in &block.stmts {
-            self.visit_stmt(stmt);
+            //add a check for return statement type matching the function return type if provided
+            let temp = self.visit_stmt(stmt);
+            if temp.is_some() {
+                //throw an error if not all returns give the same return type
+                if ret_type.is_none() {
+                    ret_type = temp;
+                } else if ret_type != temp {
+                    self.err.type_check_error(
+                        false,
+                        "Return type of statement does not match the return type of prior return statements".to_owned(),
+                        &stmt.loc().clone().map(|l| l.line_col),
```

Comment:
You can add more information to this error message:
- the `loc` of expected return type
- the expected return type
- the `loc` with the conflicting return type
- the conflicting return type